File: leetcode/leetcode_1755.cpp
/**
 * @file leetcode_1755.cpp
 * @author Cattle_Horse (CattleHorse@duck.com)
 * @brief https://leetcode.cn/problems/closest-subsequence-sum/
 * @version 0.1
 * @date 2024-06-05
 *
 * @copyright Copyright (c) 2024
 *
 */

#include <algorithm>
#include <bit>
#include <cstdint>
#include <cstdlib>
#include <functional>
#include <iterator>
#include <limits>
#include <vector>

template<typename ValueType>
class Combinations {
    using size_type = uint32_t;
    using operator_type = std::function<ValueType(const ValueType&, const ValueType&)>;

public:
    template<typename RandomAccessIterator>
    static std::vector<ValueType> get_subsequence_set(RandomAccessIterator first, RandomAccessIterator last, operator_type op = std::plus<ValueType>(), const ValueType& zero_value = ValueType()) {
        size_type n = std::distance(first, last), S = 1 << n;
        std::vector<ValueType> subsequence_set(S, zero_value);
        for (size_type s{1u}; s < S; ++s) {
            subsequence_set[s] = op(subsequence_set[s & (s - 1)], *(first + std::countr_zero(s)));
        }
        return subsequence_set;
    }
    template<typename RandomAccessIterator>
    static std::vector<std::vector<ValueType>> get_subsequence_set_count(RandomAccessIterator first, RandomAccessIterator last, operator_type op = std::plus<ValueType>(), const ValueType& zero_value = ValueType()) {
        size_type n = std::distance(first, last), S = 1 << n;
        std::vector<std::vector<ValueType>> result(n + 1);
        result[0].push_back(zero_value);
        std::vector<ValueType> temp(S, zero_value);
        for (size_type s{1u}; s < S; ++s) {
            temp[s] = op(temp[s & (s - 1)], *(first + std::countr_zero(s)));
            result[std::popcount(s)].push_back(temp[s]);
        }
        return result;
    }
};

class Solution {
public:
    int minAbsDifference(std::vector<int>& nums, int goal) {
        int m = nums.size(), leftSize = m >> 1, rightSize = m - leftSize;
        // 折半, 左半选一些, 记录所有情况, 排序
        // 枚举右半, 二分(或双指针)找到最接近goal的和
        std::vector<int> leftSum{Combinations<int>::get_subsequence_set(nums.begin(), nums.begin() + leftSize)};
        std::vector<int> rightSum{Combinations<int>::get_subsequence_set(nums.begin() + leftSize, nums.end())};
        std::sort(leftSum.begin(), leftSum.end());
        std::sort(rightSum.begin(), rightSum.end());
        int ans{std::numeric_limits<int>::max()};
        int l = 0, r = rightSum.size() - 1;
        while (l < leftSum.size() && r >= 0) {
            int s = leftSum[l] + rightSum[r];
            ans = std::min(ans, std::abs(goal - s));
            if (s > goal) {
                --r;
            } else {
                ++l;
            }
        }
        return ans;
    }
};

File: leetcode/leetcode_1755.py
"""
https://leetcode.cn/problems/closest-subsequence-sum/description/
1755. 最接近目标值的子序列和
折半搜索
"""

import functools
import math
import operator
from typing import List


def tail_zero_count(x: int) -> int:
    """计算x的尾部0的个数"""
    s = 0
    while x & 1 == 0:
        s += 1
        x >>= 1
    return s


def subsequence_list(arr: list, op, initial) -> list:
    """获得arr的子集运算结果表"""
    n = len(arr)
    S = 1 << n
    res = [initial] * S
    for s in range(1, S):
        res[s] = op(res[s & (s - 1)], arr[tail_zero_count(s)])
    return res


def subsequence_count_list(arr: list, op, initial) -> list[list]:
    """获得arr的子集运算结果表并按二进制1的个数分组"""
    n = len(arr)
    res = [[] for _ in range(n + 1)]
    res[0].append(initial)
    S = 1 << n
    temp = [initial] * S
    for s in range(1, S):
        temp[s] = op(temp[s & (s - 1)], arr[tail_zero_count(s)])
        res[s.bit_count()].append(temp[s])
    return res


class Solution:
    def minAbsDifference(self, nums: List[int], goal: int) -> int:
        n = len(nums)
        left_size = n >> 1
        sub_list = functools.partial(subsequence_list, op=operator.add, initial=0)
        left_sum = sub_list(nums[:left_size])
        right_sum = sub_list(nums[left_size:])
        left_sum.sort()
        right_sum.sort()
        ans = math.inf
        l, r = 0, len(right_sum) - 1
        while l < len(left_sum) and r >= 0:
            s = left_sum[l] + right_sum[r]
            ans = min(ans, abs(goal - s))
            if s > goal:
                r -= 1
            else:
                l += 1
        return ans
